NativeAOT implement warning as errors as a compiler feature

vitek-karas opened the pull request:
Both roslyn and trimmer implement "warn as error" as a compiler feature - specifically the `TreatWarningsAsErrors` property only affects tools which explicitly react to it. There's a related MSBuild command line option which will affect all warnings, but that's not the one VS sets from the UI and what most customers use. See the comments in https://github.com/dotnet/runtime/issues/94178.

This implements the same behavior as the trimmer. It adds 3 new command line options to `ilc` and enable "warn as error" globally and to enable/disable specific warning codes. Note that `illink` implements a more complex command line parsing where order of parameters matters and sometimes the later ones can override the earlier ones. This was trying to emulate `csc` behavior. For `ilc` I don't see a reason to emulate that because running `ilc` on its own is not supported, only going through the SDK, which will never pass the new parameters multiple times.

For testing this enables the existing trimmer tests for this functionality. This uncovered a small issue in substitution parser, which is also fixed here. For simplicity the test infra emulates the `illink` command line behavior over `ilc` (which is easy as they're very similar).

Open question:
Trimmer introduces `ILLinkTreatWarningsAsErrors` property which defaults to `TreatWarningsAsErrors` but can be overriden and it specifically applies on to `illink`. I didn't introduce a new one for `ilc` as I don't see much value in doing so, but I may be missing something. @sbomer - do you think we should add `IlcTreatWarningsAsErrors`?

Update:
We agreed to add `IlcTreatWarningsAsErrors`.

vitek-karas committed:
Implement warnings as errors in ILC

vitek-karas committed:
msbuild integration

MichalStrehovsky commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -268,6 +268,9 @@ The .NET Foundation licenses this file to you under the MIT license.
       <IlcArg Condition="$(IlcDumpIL) == 'true'" Include="--ildump:$(NativeIntermediateOutputPath)%(ManagedBinary.Filename).il" />
       <IlcArg Condition="$(NoWarn) != ''" Include='--nowarn:"$([MSBuild]::Escape($(NoWarn)).Replace(`%0A`, ``).Replace(`%0D`, ``))"' />
+      <IlcArg Condition="$(TreatWarningsAsErrors) == 'true'" Include="--warnaserror" />
+      <IlcArg Condition="$(WarningsAsErrors) != ''" Include='--enable-warnaserror:"$([MSBuild]::Escape($(WarningsAsErrors)).Replace(`%0A`, ``).Replace(`%0D`, ``))"' />
```
Can we keep the style consistent? Most of these arguments are in the style of `nowarn`, `notrimwarn`, etc. Only a couple arguments that we share with crossgen2 have hyphens and spell out full words (crossgen2 is not consistent about the hyphen style either)

Can this be `--warnaserr` and `--nowarnaserr`? (I would prefer `warnaserror`/`nowarnaserror` but I fully expect that to be an uphill battle with System.CommandLine we don't need to fight.)

vitek-karas replied:
I can definitely change the names, but as you mention, it would be hard to mimic what roslyn/illink does due to the command line parser in use.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Trimming.Tests/TestCasesRunner/TrimmingArgumentBuilder.cs:
```diff
@@ -147,6 +147,7 @@ public virtual void AddStripLinkAttributes (bool stripLinkAttributes)
 
 		public virtual void AddSubstitutions (string file)
 		{
+			Options.SubstitutionFiles.Add (file);
```
Is this for #88647? Is it intentionally part of this PR?

vitek-karas replied:
I guess it would fix that bug as well (at least partially). 
It is intentionally part of this PR - the tests for warn-as-error in illink use substitutions files to generate the warnings (don't ask my why, I don't know why we chose to do that). I wanted to reuse the tests to get a 1:1 validation against trimmer behavior. And this does that. If it fixes other things... why not :-)

I do agree that it's a but unfortunate, but it felt easier than coming up with a separate test for this.

MichalStrehovsky reviewed (commented):
Looks like tests will need some warning disabling since the warnings now trigger a build failure.

vitek-karas committed:
PR feedback and fixes

Renames the command line arguments as per PR feedback

Adds suppressions to smoke tests as necessary

MichalStrehovsky commented:
Thank you for doing the "nice" suppressions in the tests (instead of just blanket disabling everything).

Looks like there's more failures outside nativeaot/SmokeTests.

We should also run the pri0 suite since I assume there will be more of these in the test tree. If you don't want to run it locally, I suggest disabling WarnAsError in the PR temporarily so that the build doesn't just stop after a couple of these (or look at warnings in some older PRs where we did `azp run runtime-nativeaot-outerloop`).

vitek-karas committed:
PR feedback and test fixes

vitek-karas commented:
I added the `IlcTreatWarningsAsErrors`.
And then I used it in tests to disable it - to avoid breaking tests if some code doesn't correctly suppress all the right warnings. Basically as before the change.
I guess we could try fixing the tests, but this felt simpler.

vitek-karas commented:
/azp run runtime-nativeaot-outerloop

vitek-karas commented:
All the failures seem to be known (The S.T.Json failure is the crash without a dump problem reported elsewhere already).
@MichalStrehovsky could you please take a look at the recent test changes if you like it?

MichalStrehovsky reviewed (approved):
Thank you! Sorry for the delay, I thought I signed off already.
